<center><img src='img/ms_logo.jpeg' width=30% height=30%></center>
<center><h1>Descriptive Statistics</h1></center>

In this lesson, we'll clean and analyze the Titanic data set, and see if descriptive statistics can tell us anything interesting about the data set.  Along the way, we'll gain hands-on experience with industry-standard tools such as the `pandas` library in python. 

#### About the Data Set.  

The Titanic sank after hitting an iceberg the night of May 31st, 1911.  1503 people died--only 705 survived (could have been 706, but some people are door hogs).  

<img src='http://4.bp.blogspot.com/-QZUM4Q23E3c/UJ7WFXdJABI/AAAAAAAAAXk/ityxfFzjpPE/w1200-h630-p-k-no-nu/titanic1.jpg' height=50% width=50%>


The Titanic dataset is often used as an introduction to data analysis, especially for those interested in machine learning. Because we know who survived and who didn't, we can use the data on passengers to explore, look for trends that affect survival rate, and maybe even make some predictions on which passengers survived by looking at their data.   Today, we'll explore the Titanic dataset and get some real-world practice with cleaning and manipulating data.  

#### Pandas--Favorite Tool of Data Scientists 

<img src='https://media.giphy.com/media/aUhEBE0T8XNHa/giphy.gif' height=25% width=25%>

For data processing in Python, you can't beat the `pandas` library.  Pandas is used for creating dataframes, which are reminiscent of tables or spreadsheets, but much more powerful and easy to use.  Pandas provides a clean, easy-to-use API for reading, manipulating, sorting, and slicing data.  

#### Learning Goals

In this exercise, we have the following goals:

1.  Use pandas to read-in and manipulate data.
1.  Explore strategies for detecting outliers and dealing with missing (NaN) values
1.  Answer questions about our data set using descriptive statistics.
1.  Use pandas to slice our dataframe into smaller dataframes based on conditional logic (for instance, all female survivors under a certain age)

**Let's get started!**

(To execute any cell, press SHIFT + ENTER or press the 'Run' button on the toolbar at the top.)

We'll start by importing pandas and setting an alias:

In [5]:
# We'll be using pandas a lot, so let's alias it to the name 'pd' to save ourselves some keystrokes
import pandas as pd

Next, we'll want to read in our dataset from the from the the 'datasets' folder in this directory.  Did you know jupyter notebooks can also use terminal commands in code cells?  Just put a '%' sign before the command and it works the same as in the terminal.  For instance, if you wanted to list all the items contained in this current directory, you would just type `%ls`!  Try it below:

In [6]:
# Type % ls and run this cell
%ls

00_Test_Installation.ipynb        05__AB Testing.ipynb
01__Descriptive Statistics.ipynb  Installation_Instructions.md
02__Data Visualization.ipynb      README.md
03__Distributions_and_CLT.ipynb   datasets/
04__Hypothesis Testing .ipynb     img/


Our data set is contained within the `datasets` folder, and is called `titanic.csv`.  You may have some experience reading data in manually using the `open` keyword in python.  With large/complex datasets, this method can get tedious very quickly.  However, `pandas` makes this a simple task!

**TASK: Read in the titanic.csv file using the pd.read_csv( ) method in pandas.**

HINT:  You'll still need to pass the method the correct arguments--in order to get this right, You'll need the documentation for this method.  You can find that [here](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html).  

Alternatively, you can just type the method name with a question mark instead of parentheses to open a method's docs right here in the notebook!

In [20]:
# Need the docs? try typing pd.read_csv? 
path = "datasets/titanic.csv"
df = pd.read_csv(path)   # store the newly created dataframe in this variable
df   # this will display the contents of df (the dataframe containing all our data, truncated for readability)

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
5              6         0       3   
6              7         0       1   
7              8         0       3   
8              9         1       3   
9             10         1       2   
10            11         1       3   
11            12         1       1   
12            13         0       3   
13            14         0       3   
14            15         0       3   
15            16         1       2   
16            17         0       3   
17            18         1       2   
18            19         0       3   
19            20         1       3   
20            21         0       2   
21            22         1       2   
22            23         1       3   
23            24         1       1   
24            25         0       3   
25            26         1       3   
26            27         0       3   
27            28         0       1   
28            29         1       3   
29            30         0       3   
..           ...       ...     ...   
861          862         0       2   
862          863         1       1   
863          864         0       3   
864          865         0       2   
865          866         1       2   
866          867         1       2   
867          868         0       1   
868          869         0       3   
869          870         1       3   
870          871         0       3   
871          872         1       1   
872          873         0       1   
873          874         0       3   
874          875         1       2   
875          876         1       3   
876          877         0       3   
877          878         0       3   
878          879         0       3   
879          880         1       1   
880          881         1       2   
881          882         0       3   
882          883         0       3   
883          884         0       2   
884          885         0       3   
885          886         0       3   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
5                                     Moran, Mr. James    male   NaN      0   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
7                       Palsson, Master. Gosta Leonard    male   2.0      3   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
12                      Saundercock, Mr. William Henry    male  20.0      0   
13                         Andersson, Mr. Anders Johan    male  39.0      1   
14                Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                                Rice, Master. Eugene    male   2.0      4   
17                        Williams, Mr. Charles Eugene    male   NaN      0   
18   Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                             Masselmani, Mrs. Fatima  female   NaN

Great! With the pandas library, importing large datasets becomes really, really easy.  We can also use it to examine the data, and manipulate the dataset very easily. Before we can make sense of the data, we should probably have a feel for what each column means.  Here's the **Data Dictionary** explaining what each column and value actually means:


<center><h2>Data Dictionary</h2></center>


| Variable | Definition                                        | Key                                            |
|----------|---------------------------------------------------|------------------------------------------------|
| Survived | Survival                                          | 0 = No, 1 = Yes                                |
| Pclass   | Ticket Class (proxy for socio-economic status)    | 1 = 1st, 2 = 2nd, 3 = 3rd                      |
| Sex      | Sex                                               |                                                |
| Age      | Age (in years)                                    |                                                |
| SibSp    | # of siblings and/or spouses also aboard          |                                                |
| Parch    | # of parents / children also aboard               |                                                |
| Ticket   | Ticket number                                     |                                                |
| Fare     | Passenger fare (how much their ticket cost)       |                                                |
| Cabin    | Cabin number                                      |                                                |
| Embarked | Port of Embarkation (where they boarded the ship) | C = Cherbourg, Q = Queenstown, S = Southampton |

Let's get the basic descriptive statistics to see what we can figure out about this dataset.  

**TASK: Run the dataframe's .describe( ) command.  **

In [21]:
# Run df.describe().  What does each cell in the table mean?
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [24]:
df[df["Sex"] == "male"].describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   577.000000  577.000000  577.000000  453.000000  577.000000   
mean    454.147314    0.188908    2.389948   30.726645    0.429809   
std     257.486139    0.391775    0.813580   14.678201    1.061811   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     222.000000    0.000000    2.000000   21.000000    0.000000   
50%     464.000000    0.000000    3.000000   29.000000    0.000000   
75%     680.000000    0.000000    3.000000   39.000000    0.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  577.000000  577.000000  
mean     0.235702   25.523893  
std      0.612294   43.138263  
min      0.000000    0.000000  
25%      0.000000    7.895800  
50%      0.000000   10.500000  
75%      0.000000   26.550000  
max      5.000000  512.329200

We can also call `.describe()` on individual columns (a single column of data is called a pandas _series_).  For instance, if we wanted to see the summary statistics on the _SibSp_ column, we would type `df["SibSp"].describe()`

See if you can answer the following questions:  

1.  How old is the oldest passenger on the titanic?
1.  How young is the youngest passenger?
1.  What is the average price paid for a ticket?
1.  How much did the most expensive ticket cost?
1.  How many passengers in the dataset are female?

Answers:

1.  80 years old
1.  5 months old
1.  \$32.20
1.  \$512.33
1.  314

In [19]:
# Use the .describe() method on the appropriate columns to answer the questions listed above.
df["Age"].describe()

count    714.000000
mean      29.699118
std       14.526497
min        0.420000
25%       20.125000
50%       28.000000
75%       38.000000
max       80.000000
Name: Age, dtype: float64

<center><h3>Slicing</h3></center>

We can also use pandas to 'slice' dataframes, just as we would with a list in python.  However, unlike list slicing in vanilla python, we can slice using conditional logic.  For instance, what if we want to examine a dataframe that only contains the passengers that survived?  Easy-- we just type:

<center>`survived_df = df[df["Survived"] == 1]`</center>

The syntax for slicing can feel a bit clunky at first, but it will become intuitive with practice.  

If you want to slice on multiple conditions, you can do that too!  Just make sure each condition is wrapped in a set of parentheses.  For instance, if we wanted to grab a dataframe filled only when men that survived, we would use:
<br>
<br>
<center>`df[(df['Survived'] == 1) & (df['Sex'] == 'male')]`</center>


**TASK: Use your knowledge of conditional slicing to answer the following questions:**

(HINT: Don't forget about the `.describe()` method!)

1.  How many men survived?
1.  What is the average age of male passengers that survived?
1.  How many female passengers under the age of 30 did not survive?
1.  What was the most expensive ticket bought by answer passenger from Cherbourg that did NOT survive?
1.  Of all surviving passengers from Southampton, how many passengers paid between \$216 and \$676.50 for their ticket? 

**Answers:**

1.  93
1.  27
1.  42
1.  \$247.52
1.  2

In [12]:
# Use your knowledge of conditional slicing to answer the questions above!
survived_df = df[df["Survived"] == 1]
survived_df

PassengerId  Survived  Pclass  \
1              2         1       1   
2              3         1       3   
3              4         1       1   
8              9         1       3   
9             10         1       2   
10            11         1       3   
11            12         1       1   
15            16         1       2   
17            18         1       2   
19            20         1       3   
21            22         1       2   
22            23         1       3   
23            24         1       1   
25            26         1       3   
28            29         1       3   
31            32         1       1   
32            33         1       3   
36            37         1       3   
39            40         1       3   
43            44         1       2   
44            45         1       3   
47            48         1       3   
52            53         1       1   
53            54         1       2   
55            56         1       1   
56            57         1       2   
58            59         1       2   
61            62         1       1   
65            66         1       3   
66            67         1       2   
..           ...       ...     ...   
809          810         1       1   
820          821         1       1   
821          822         1       3   
823          824         1       3   
827          828         1       2   
828          829         1       3   
829          830         1       1   
830          831         1       3   
831          832         1       2   
835          836         1       1   
838          839         1       3   
839          840         1       1   
842          843         1       1   
849          850         1       1   
853          854         1       1   
855          856         1       3   
856          857         1       1   
857          858         1       1   
858          859         1       3   
862          863         1       1   
865          866         1       2   
866          867         1       2   
869          870         1       3   
871          872         1       1   
874          875         1       2   
875          876         1       3   
879          880         1       1   
880          881         1       2   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex    Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.00      1   
2                               Heikkinen, Miss. Laina  female  26.00      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.00      1   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.00      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.00      1   
10                     Sandstrom, Miss. Marguerite Rut  female   4.00      1   
11                            Bonnell, Miss. Elizabeth  female  58.00      0   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.00      0   
17                        Williams, Mr. Charles Eugene    male    NaN      0   
19                             Masselmani, Mrs. Fatima  female    NaN      0   
21                               Beesley, Mr. Lawrence    male  34.00      0   
22                         McGowan, Miss. Anna "Annie"  female  15.00      0   
23                        Sloper, Mr. William Thompson    male  28.00      0   
25   Asplund, Mrs. Carl Oscar (Selma Augusta Emilia...  female  38.00      1   
28                       O'Dwyer, Miss. Ellen "Nellie"  female    NaN      0   
31      Spencer, Mrs. William Augustus (Marie Eugenie)  female    NaN      1   
32                            Glynn, Miss. Mary Agatha  female    NaN      0   
36                                    Mamee, Mr. Hanna    male    NaN      0   
39                         Nicola-Yarred, Miss. Jamila  female  14.00      1   
43            Laroche, Miss. Simonne Marie Anne 

In [25]:
males_survived = df[(df['Survived'] == 1) & (df['Sex'] == 'male')]
males_survived
males_survived["Age"].describe()

count    93.000000
mean     27.276022
std      16.504803
min       0.420000
25%      18.000000
50%      28.000000
75%      36.000000
max      80.000000
Name: Age, dtype: float64

In [26]:
died = df[df["Survived"] == 0]
died.describe()

PassengerId  Survived      Pclass         Age       SibSp       Parch  \
count   549.000000     549.0  549.000000  424.000000  549.000000  549.000000   
mean    447.016393       0.0    2.531876   30.626179    0.553734    0.329690   
std     260.640469       0.0    0.735805   14.172110    1.288399    0.823166   
min       1.000000       0.0    1.000000    1.000000    0.000000    0.000000   
25%     211.000000       0.0    2.000000   21.000000    0.000000    0.000000   
50%     455.000000       0.0    3.000000   28.000000    0.000000    0.000000   
75%     675.000000       0.0    3.000000   39.000000    1.000000    0.000000   
max     891.000000       0.0    3.000000   74.000000    8.000000    6.000000   

             Fare  
count  549.000000  
mean    22.117887  
std     31.388207  
min      0.000000  
25%      7.854200  
50%     10.500000  
75%     26.000000  
max    263.000000

In [29]:
females_dead_u30 = df[(df["Sex"] == "female") & (df["Age"] < 30) & (df["Survived"] == 0)]
females_dead_u30.describe()

PassengerId  Survived     Pclass        Age      SibSp      Parch  \
count    42.000000      42.0  42.000000  42.000000  42.000000  42.000000   
mean    393.214286       0.0   2.833333  17.130952   1.238095   0.857143   
std     265.052368       0.0   0.489732   8.400827   1.445085   1.001741   
min      15.000000       0.0   1.000000   2.000000   0.000000   0.000000   
25%     127.000000       0.0   3.000000   9.250000   0.000000   0.000000   
50%     404.000000       0.0   3.000000  18.000000   1.000000   0.500000   
75%     605.500000       0.0   3.000000  24.000000   2.000000   2.000000   
max     883.000000       0.0   3.000000  29.000000   5.000000   4.000000   

             Fare  
count   42.000000  
mean    24.095933  
std     30.443291  
min      6.750000  
25%      9.825000  
50%     14.852050  
75%     27.425000  
max    151.550000

In [37]:
# What was the most expensive ticket bought by answer passenger from Cherbourg that did NOT survive?
cherbourg = df[(df["Embarked"] == "C") & (df["Survived"] == 0)]
cherbourg.describe()

PassengerId  Survived     Pclass        Age      SibSp      Parch  \
count    75.000000      75.0  75.000000  51.000000  75.000000  75.000000   
mean    409.853333       0.0   2.200000  33.666667   0.253333   0.253333   
std     264.222556       0.0   0.929981  14.990553   0.467695   0.547558   
min      27.000000       0.0   1.000000   9.000000   0.000000   0.000000   
25%     148.500000       0.0   1.000000  23.750000   0.000000   0.000000   
50%     378.000000       0.0   3.000000  30.000000   0.000000   0.000000   
75%     610.000000       0.0   3.000000  42.500000   0.000000   0.000000   
max     860.000000       0.0   3.000000  71.000000   2.000000   2.000000   

             Fare  
count   75.000000  
mean    35.443335  
std     49.617218  
min      4.012500  
25%      7.229200  
50%     14.458300  
75%     32.675000  
max    247.520800

In [41]:
# Of all surviving passengers from Southampton, how many passengers paid between \$216 and \$676.50 for their ticket? 
specific_passengers = df[(df["Survived"] == 1) & (df["Embarked"] == "S") & (df["Fare"] >= 216) & (df["Fare"] <= 676.5)]
specific_passengers.describe()

PassengerId  Survived  Pclass        Age  SibSp  Parch   Fare
count     2.000000       2.0     2.0   2.000000    2.0    2.0    2.0
mean    215.500000       1.0     1.0  23.500000    3.0    2.0  263.0
std     178.898016       0.0     0.0   0.707107    0.0    0.0    0.0
min      89.000000       1.0     1.0  23.000000    3.0    2.0  263.0
25%     152.250000       1.0     1.0  23.250000    3.0    2.0  263.0
50%     215.500000       1.0     1.0  23.500000    3.0    2.0  263.0
75%     278.750000       1.0     1.0  23.750000    3.0    2.0  263.0
max     342.000000       1.0     1.0  24.000000    3.0    2.0  263.0